### ISSUE TITLE
Tutorial 4: Continuous control with SAC

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#310]
Child of #303 (tutorial series). Add `docs/tutorials/04-sac-continuous.md`: SAC on PendulumSwingUp — Box action spaces, tanh action squashing, automatic entropy tuning, twin critics, Polyak target updates.

**Requirements (match tutorials 1–2):**
- Single Markdown file with complete copy-paste code.
- Code CI-enforced as a doc-test via `src/tutorials.rs` (`tutorial_04_sac` submodule).
- Builds on tutorial 3's off-policy replay machinery; adds the continuous-action pieces.
- Voice matches GETTING_STARTED.md; explain *why*.
- Update `docs/tutorials/README.md` status row.
- Full test suite + clippy -D warnings + fmt + doc build clean.

Reference the packaged `train_sac` example.

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating this issue per user request. Added below: full API surface, doc-test wrapper pattern, CI-friendly step budget, concrete acceptance criteria, scope guard, implementation guidance, affected files, and test plan. Tutorials 1–3 are all on `origin/main` (tutorial 3 merged today as PR #319).

---

### Context: build base is current

`src/tutorials.rs` now has three `pub mod tutorial_0N_*` entries (1–3). Tutorial 4 needs a fourth entry following the same two-line pattern. The `docs/tutorials/README.md` status row for #310 already reads "🚧 Planned" — change it to "✅ Landed".

---

### API surface

SAC has the simplest trainer call-site of the three off-policy trainers — all network construction is internal. The only imports the tutorial needs:

```rust
use burn::backend::{Autodiff, NdArray};
use thrust_rl::{
    env::{Environment, games::pendulum::PendulumSwingUp},
    train::sac::{SacConfig, SacTrainer},
};
type Backend = Autodiff<NdArray<f32>>;
```

**`SacConfig`** — builder-style setters (every field has a setter; defaults are Haarnoja 2018 v2 paper values):

| Setter | What it controls |
|---|---|
| `buffer_capacity(usize)` | replay FIFO ring capacity |
| `min_buffer_size(usize)` | transitions required before first gradient step |
| `learning_starts(usize)` | steps of pure-random warmup before actor is used |
| `batch_size(usize)` | minibatch size per gradient step |
| `hidden_dim(usize)` / `num_hidden_layers(usize)` | actor + critic network width/depth |
| `actor_lr(f64)` / `critic_lr(f64)` / `alpha_lr(f64)` | three independent Adam learning rates |
| `gamma(f64)` | discount factor |
| `tau(f64)` | Polyak blend coefficient (default `0.005`; applied every gradient step) |
| `auto_alpha(bool)` / `init_alpha(f32)` / `target_entropy(f32)` | entropy temperature tuning |
| `max_grad_norm(f64)` | optional global gradient-norm clip |
| `seed(u64)` | reproducibility (actor, twin critics, replay RNG all derived from this) |

`SacConfig::new().validate()` is called inside `SacTrainer::new` — no explicit validate call needed.

**`SacTrainer::new(config, obs_dim, action_dim, device) -> Result<SacTrainer<B>>`** — constructs everything: twin online critics `q1`/`q2`, twin targets `q1_target`/`q2_target`, `SacActor`, `LogAlpha`, three Adam optimizers, and `ContinuousReplayBuffer`.

**The interaction loop** (simpler than DQN — no explicit target-sync closure, no argmax closure):

```text
action = trainer.select_action(&obs)                    // warmup: uniform (-1,1); after: actor sample
env.step(scale_action(&action))                         // rescale tanh output by MAX_TORQUE = 2.0
trainer.buffer_mut().push(&obs, &action, r, &next, done)
trainer.increment_env_step()
trainer.train()?                                        // no-op until min_buffer_size; returns Option<SacStepStats>
if done { trainer.increment_episodes(1); env.reset(); }
```

`select_action` handles the warmup/policy switch internally via `in_warmup()` — no ε-greedy closure needed. `train()` runs `gradient_steps_per_env_step` full SAC updates (critics → actor → alpha → Polyak targets) in one call.

**`PendulumSwingUp`** facts the tutorial must state:
- `obs_dim = 3`: `[cos θ, sin θ, θ̇]` — angle encoded as (cos, sin) so the observation is continuous across the ±π wrap
- `action_dim = 1`: torque, env clamps to `[-2.0, 2.0]`; actor outputs `(-1, 1)`, so multiply by `MAX_TORQUE = 2.0` before `env.step()`
- Episode length: 200 steps, then `truncated = true`; `terminated` is never set
- Reward: always ≤ 0; near-optimal ~-150; random policy ~-1200...-1600
- Construction: `PendulumSwingUp::with_seed(SEED)` for reproducibility

---

### Doc-test wrapper pattern (mirrors tutorial 3 exactly)

Two edits to `src/tutorials.rs`:

**1.** Add the module entry after `tutorial_03_dqn`:

```rust
/// Tutorial 4 — Continuous control with SAC.
///
/// See the rendered Markdown at
/// [`docs/tutorials/04-sac-continuous.md`](https://github.com/rjwalters/thrust/blob/main/docs/tutorials/04-sac-continuous.md).
#[doc = include_str!("../docs/tutorials/04-sac-continuous.md")]
pub mod tutorial_04_sac {}
```

**2.** Update the module-level "Landed tutorials" doc comment to add a bullet for `tutorial_04_sac`.

The Markdown code block must be a plain `rust` block (no `# extern crate`, no `#[test]` attribute, no `fn main()`). Rustdoc wraps every doc-test in an implicit `fn main()` block. The `type Backend = ...` alias goes at the top of the block; `let device = Default::default()` is the first statement.

---

### CI-friendly step budget

The doc-test runs on every `cargo test --features training`. With the `NdArray<f32>` CPU backend and tiny networks, the loop must complete in well under 30 seconds:

| Parameter | CI value | Full-run value (reference: `train_sac` example) |
|---|---|---|
| `TOTAL_TIMESTEPS` | 4_096 | 30_000 |
| `buffer_capacity` | 4_096 | 50_000 |
| `min_buffer_size` | 512 | 1_000 |
| `learning_starts` | 512 | 1_000 |
| `batch_size` | 64 | 256 |
| `hidden_dim` | 32 | 256 |
| `num_hidden_layers` | 2 | 2 |

Terminal assert: `assert!(trainer.total_env_steps() >= TOTAL_TIMESTEPS);` — identical form to tutorial 3. Do not assert a reward threshold in the doc-test.

---

### Scope guard

Out of scope for this issue:
- No new packaged example. `examples/games/pendulum/train_sac.rs` is the full-run reference; the tutorial points to it.
- No evaluation loop beyond confirming the loop ran to `TOTAL_TIMESTEPS`.
- No multi-env or vectorized SAC.
- No TD3, DDPG, or other continuous-action variants.
- No changes to any `src/` files other than `src/tutorials.rs` (the SAC trainer/config API is already complete).

---

### Implementation guidance

**Structure closely follows tutorial 3.** Suggested section order:

1. Summary callout block ("You will…", prerequisites: Tutorial 3, time: ~20 min)
2. Why continuous control is different — Box action spaces, why DQN's argmax over actions doesn't apply, what a policy distribution over ℝ looks like
3. The problem: PendulumSwingUp — physics, observation encoding, reward signal, why return is always ≤ 0
4. SAC's four pieces — stochastic actor, twin critics, entropy temperature, Polyak targets. **Frame as extending tutorial 3's replay + target machinery**: replay buffer and Polyak are already familiar; SAC adds the continuous-action actor and entropy regularization.
5. Tanh action squashing — why it constrains actions to `(-1, 1)`, the change-of-variables log-prob correction (cite Haarnoja App. C), the `MAX_TORQUE` rescaling step
6. Automatic entropy tuning — contrast with DQN's ε schedule; the `target_entropy = -action_dim` heuristic; what "temperature" means geometrically
7. Twin critics — double-Q underestimation fix; relate to Double-DQN (same family of insight: decouple selection from evaluation)
8. The trainer call-site is simpler than DQN — call out explicitly that no closure or explicit target-sync call is needed
9. The code (CI-sized loop, complete copy-paste)
10. Reading the learning curve — point to `train_sac` + `CURVE_CSV`; explain the negative-return axis; note that Pendulum curves should NOT be overlaid on CartPole DQN/PPO curves (different reward scales)
11. When to prefer SAC over DQN/PPO — continuous vs. discrete; entropy regularization vs. ε-greedy exploration
12. Try it yourself — disable auto-alpha, swap `hidden_dim`, point to `MountainCarContinuous` as a stretch env

Voice contract: explain *why* at every step. Do not re-explain replay buffers or Polyak from scratch — reference them as "the machinery from tutorial 3" and link.

---

### Acceptance Criteria

- [ ] `docs/tutorials/04-sac-continuous.md` exists as a single Markdown file with complete, copy-paste code block.
- [ ] `src/tutorials.rs` has a `pub mod tutorial_04_sac {}` entry with `#[doc = include_str!("../docs/tutorials/04-sac-continuous.md")]`.
- [ ] Module-level doc comment in `src/tutorials.rs` lists `tutorial_04_sac` in the "Landed tutorials" section.
- [ ] `docs/tutorials/README.md` status row for tutorial 4 updated to "✅ Landed".
- [ ] `cargo test --features training -- tutorial_04` passes (doc-test compiles and completes).
- [ ] Doc-test `TOTAL_TIMESTEPS` ≤ 8_192, `hidden_dim` ≤ 64; terminal assert is `assert!(trainer.total_env_steps() >= TOTAL_TIMESTEPS)` only — no convergence bar.
- [ ] Tutorial references `examples/games/pendulum/train_sac.rs` as the full-run counterpart with `CURVE_CSV`.
- [ ] `cargo clippy --features training -- -D warnings` clean.
- [ ] `cargo fmt --check` clean.
- [ ] `cargo doc --features training --no-deps` builds without errors or warnings.

---

## Affected Files

- `docs/tutorials/04-sac-continuous.md` — (new) create tutorial Markdown; single source of truth for prose + CI-enforced code block
- `src/tutorials.rs` — add `pub mod tutorial_04_sac {}` entry; update module-level "Landed tutorials" doc listing
- `docs/tutorials/README.md` — update tutorial 4 status row from "🚧 Planned" to "✅ Landed"

**Curator pre-flight**: `src/tutorials.rs` and `docs/tutorials/README.md` confirmed present on `origin/main`. `docs/tutorials/04-sac-continuous.md` is a genuinely new file (not an uncommitted working-tree artifact) — the builder creates it from scratch.

---

## Test Plan

- [ ] `cargo test --features training -- tutorial_04` passes and completes in < 30 s on CI hardware
- [ ] `cargo clippy --features training -- -D warnings` reports zero warnings
- [ ] `cargo fmt --check` passes
- [ ] `cargo doc --features training --no-deps` builds without warnings
- [ ] Code block is idiomatic: no `extern crate`, no `#[test]`, no `fn main()`, no stray `use` outside the block
- [ ] Tutorial voice check: each section answers "why" not just "what"; replay/target concepts reference tutorial 3 by name
- [ ] Learning-curve section explicitly notes that Pendulum returns are negative and should not be overlaid on CartPole curves
